clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
+
+        if (consumerThread != null && consumerThread.isAlive()) {
+            rebalanceConsumer.sendRequest(Duration.ofMillis(remainingTimeAtLeastZero(timeoutMs, elapsedMs)),
+                                          RebalanceKafkaConsumer.ConsumerRequest.POLL,
+                             null,
+                                          new DefaultTaskCompletionCallback());
+        }
+    }
+
+    private RebalanceKafkaConsumer.RequestResult pollForResults(final long timeoutMs, final long now) {
+        long elapsed = time.milliseconds() - now;
+        boolean condition = remainingTimeAtLeastZero(timeoutMs, elapsed) != 0;
+        while (result == null && condition) {
+            try {
+                Thread.sleep(retryBackoffMs);
+            } catch (InterruptedException exc) { }
+            elapsed = time.milliseconds() - now;
+            condition = remainingTimeAtLeastZero(timeoutMs, elapsed) != 0;
+        }
+        return result;
+    }
+
+    private ConsumerRecords<K, V>  mergeRecords(final ConsumerRecords<K, V> records1, final ConsumerRecords<K, V> records2) {
+        final HashMap<TopicPartition, List<ConsumerRecord<K, V>>> map = new HashMap<>();
+        for (final TopicPartition partition : records1.partitions()) {
+            map.put(partition, records1.records(partition));
+        }
+        for (final TopicPartition partition : records2.partitions()) {
+            if (!map.containsKey(partition)) {
+                map.put(partition, records2.records(partition));
+            } else {
+                map.get(partition).addAll(records2.records(partition));
+            }
+        }
+        return new ConsumerRecords<>(map);
+    }
+
+    private ConsumerRecords<K, V> processRecords(final long timeoutMs,
+                                                 final long checkRebalanceStart,
+                                                 final Map<TopicPartition, List<ConsumerRecord<K, V>>> records) {
+        // before returning the fetched records, we can send off the next round of fetches
+        // and avoid block waiting for their responses to enable pipelining while the user
+        // is handling the fetched records.
+        //
+        // NOTE: since the consumed position has already been updated, we must not allow
+        // wakeups or any other errors to be triggered prior to returning the fetched records.
+        if (fetcher.sendFetches() > 0 || client.hasPendingRequests()) {
+            client.pollNoWakeup();
+        }
+
+        if (consumerThread != null && consumerThread.isAlive()) {
+            ConsumerRecords<K, V> offsetLagRecords =
+                    result == null ? (ConsumerRecords<K, V>) pollForResults(timeoutMs, checkRebalanceStart).value
+                            : (ConsumerRecords<K, V>) result.value;
+            if (offsetLagRecords == null) {
+                return this.interceptors.onConsume(new ConsumerRecords<>(records));
```

Comment:
nit: extract the call to this.interceptors.onConsume(new ConsumerRecords<>(records)) above line 1258 - its result would always be used